[8.0] sale_commission_formula

dcorio opened the pull request:
module developed to satisfy a customer's specific need.
it should be generic enough for the inclusion in this OCA's project.

Commit:
[ADD] sale_commission_formula, initial rev

Commit:
[IMP] formula computation

Commit:
[IMP] formula attrs simplification

Commit:
Merge branch '8.0' of github.com:OCA/commission into 8.0-commission-formula

Commit:
[IMP] v8 api porting

Commit:
[IMP] add README

dcorio commented:
I'd like to have your opinion and suggestions about this module.
thanks in advance.

nhomar commented:
@dcorio sorry closed by mistake :-(

Commit:
[FIX] new line EOF

Commit:
[FIX] wrong param

pedrobaeza commented on sale_commission_formula/models/account_invoice.py:
```diff
@@ -0,0 +1,40 @@
+            self.amount = 0.0
+            formula = self.commission.formula
+            results = {'line': self.invoice_line}
+            exec(formula) in results
```
Maybe you should use safe_eval or similar to avoid the injection of malicious code

dcorio replied:
Thanks Pedro, i'll change my code accordingly. I assumed exec was secure enough because i saw Odoo using it elsewhere.

pedrobaeza commented on sale_commission_formula/models/sale_commission.py:
```diff
@@ -0,0 +1,33 @@
+    formula = fields.Text(
+        'Formula',
+        default="""# Use 'line' to access all the invoice line attributes\n
+result = 0 # use 'result' to return the commission amount""")
```
Don't use triple quota, but simple ones enclosed on each line

gurneyalex committed:
Update travis configuration

use sudo: false to enable container build
use cache: pip to cache pip packages

pedrobaeza commented:
Hi, @dcorio,

I have checked your module and it works well, but there are a few issues here to solve:
- You are using the same formula for sale orders and for invoices, but passing line as a different object (account.invoice.line or sale.order.line), so there's the possibility to write a formula that works in one place but not in the other. For solving this, we can:
  - Define 2 formulas, one for the sale order and one for the account invoice.
  - Document in the same field how can we compare in the formula to see if we are at sale order or at account invoice (`if line._name == 'sale.order.line': ...`).
- The data to compute the commission can be only in the sale order, so we need a way to indicate that the amount mustn't be recomputed on account invoices.

rafaelbn commented:
Hi @dcorio as @pedrobaeza says we must fix the formula's problem as commissions could be in sales.order or more often in invoices. Could tell us something about this? I will test it. Thanks

dcorio commented:
@rafaelbn @pedrobaeza thanks for the feedback and suggestions. i'll get back to you in the very few next days.

sbidoul committed:
[ADD] addons table in README.md

sbidoul committed:
[UPD] prefix versions with 8.0

Commit:
[FIX] use safe_eval

Commit:
[IMP] inline instructions

dcorio commented:
@rafaelbn @pedrobaeza i finally found the time to fix those things on this module. i didn't split the formula field in two since i wouldn't make sense considering the fact that one should not apply different commission rates on SO and invoice. So i decided to better explain how to configure the commission. Right to the field you now have an explanation and a table with the most used fields for the two modules. The default formula already provides the correct structure.

I also converted the exec() call into a safe_eval one.

pedrobaeza commented:
Thanks for the changes. You should differentiate the icon from the generic one. You can edit the SVG with Inkscape and add a text with some formula scheme like "x = y + z" to represent the goal of this module. Tests are also welcome.

Commit:
[FIX] dedicated icon

Commit:
[FIX] indentation

Commit:
[FIX] safe_eval direct call

dcorio commented:
thank you very much @pedrobaeza for the review. could you please have a look at the code now?

dcorio commented:
PS: working on the tests

pedrobaeza commented:
Thanks, :+1: waiting for the tests

Commit:
[ADD] tests

Commit:
[IMP] tests comments

tafaRU commented:
@dcorio, thanks for this module!

Also please adapt the README file to https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst and update the version to https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

Commit:
[FIX] "sale" journal type on making agent invoice

On generate agent invoice you can only select journal of type "sale". This can cause the numbers of sales invoices not correlated.
IMHO, agent invoices must be of type purchase and this solves it, allowing to select journals of type purchase.

pedrobaeza committed:
[ADD] contract_commission

=======================
Commission in contracts
=======================

This module allows to get default commissions for partner in the recurring
invoices generated from contracts.

Installation
============

This module is auto-installed when you have *account_analytic_analysis* and
*sale_commission* modules installed.

Usage
=====

When invoicing contracts, the agents assigned to the partner of the contract
will now be assigned to the generated invoices.

mbcosta committed:
Created new field to specify the commission creation according to the state of the invoice

mbcosta committed:
Fixed the demo datas and create tests.

oihane committed:
[IMP] sale_commission: base amount for computation selectable
You can now select if commission is calculated from subtotal
or from margin

* added python tests

OCA-git-bot committed:
[UPD] addons table in README.md

hurrinico commented:
Hi @dcorio ,
Thanks for the module
Could i ask you to rebase your branch?
Thank You in advance

ivantodorovich commented:
:+1: waiting for this merge!

pedrobaeza commented:
@dcorio, please update latest comments about README file and I'll merge this one.

dcorio commented:
@pedrobaeza i'll do the update this evening. sorry for the delay but i'm quite busy these days.

Commit:
[FIX] Fix singleton error

pedrobaeza commented:
It seems the rebase have failed. @dcorio, can you please check it and also squash a bit your commits to a few?

dcorio commented:
@pedrobaeza noticed that thanks. going to fix it

hurrinico commented:
@dcorio @pedrobaeza to fix this problem i have done: https://github.com/hurrinico/commission/tree/add_sale_commission_formula if you want i can do another PR in order to solve the problem

dcorio commented:
@hurrinico sure! this will save time

hurrinico commented:
@dcorio @pedrobaeza I've created this https://github.com/OCA/commission/pull/48

eLBati commented:
@hurrinico @dcorio can this be closed?

dcorio commented:
@eLBati yes, we can continue on https://github.com/OCA/commission/pull/48
